flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1256,49 +1307,82 @@ public void testProcessWithUnAvailableOutput() throws Exception {
 
             final RunnableWithException completeFutureTask =
                     () -> {
-                        Thread.sleep(sleepTime + 1);
                         assertEquals(1, inputProcessor.currentNumProcessCalls);
                         assertTrue(task.mailboxProcessor.isDefaultActionUnavailable());
                         environment.getWriter(1).getAvailableFuture().complete(null);
                     };
 
+            // Make sure WaitingThread is started after Task starts processing.
             executor.submit(
-                    () -> {
-                        executor.submit(
-                                completeFutureTask,
-                                "This task will complete the future to resume process input action.");
-                    },
+                    () ->
+                            new WaitingThread(
+                                            executor,
+                                            completeFutureTask,
+                                            sleepTimeInsideMail,
+                                            sleepTimeOutsideMail)
+                                    .start(),
                     "This task will submit another task to execute after processing input once.");
 
+            long startTs = System.currentTimeMillis();
             TaskIOMetricGroup ioMetricGroup =
                     task.getEnvironment().getMetricGroup().getIOMetricGroup();
             task.invoke();
+            long totalDuration = System.currentTimeMillis() - startTs;
             assertThat(
                     ioMetricGroup.getBackPressuredTimePerSecond().getCount(),
-                    Matchers.greaterThanOrEqualTo(sleepTime));
+                    Matchers.greaterThanOrEqualTo(sleepTimeOutsideMail));
+            assertThat(
+                    ioMetricGroup.getBackPressuredTimePerSecond().getCount(),
+                    Matchers.lessThanOrEqualTo(totalDuration - sleepTimeInsideMail));
```

Comment:
How reliable is this test?
I guess since we compare to the actually measured `totalDuration`, it's stable also in situations like GC pauses.

But is it also effectively testing? What happens if `totalDuration` is much larger than any of the sleep times?
(I just want to avoid any test instabilities by explicitly going over the cases)

Reply:
The test is failing without the fix. `WaitingThread` gives us 2 guarantees. It will wait X ms before submitting a mailbox action to unblock task, and it will sleep Y ms inside that mailbox action. This directly translates that we will have 
```
idleTime > X // that's first assertion in the test, it was being tested on master before my change
busyTime > Y
```
On the other hand we know 
```
totalDuration > busyTime + idleTime
```
so...
```
totalDuration - idleTime > busyTime > Y
totalDuration - Y > idleTime  // that's the new assertion in the test
```
So as long as we can assume elapsed time measured around `Thread.sleep(X)` by java yields at least `X`, the test should be stable. I've also run in a loop couple of hundreds times.

Reply:
Okay, I just wanted to double check (after recent test instabilities).